### DISCUSSION TITLE
Multi-tenancy via path prefix?

### DISCUSSION THREAD

**Discussion by pboling** [janko/rodauth-rails#257] [Q&A] [answered]
```
prefix "/api/auth/:tenant_id"
# or alternatively...
prefix "/api/auth/:path_key"
```

Has anyone gotten something like this to work?

Such that the routes end up as:
```
/api/auth/123/create-account
# or alternatively...
/api/auth/animal-farm/create-account
```

If so, how?

**Comment by pboling** [comment]
If I set:
```
class RodauthMain < Rodauth::Rails::Auth
  configure do
    prefix "/api/auth/:path_key"
    # ...
  end
end
```
In Rails console:
```
>> RodauthApp.rodauth.create_account_path
=> "/api/auth/:path_key/create-account"
```
So perhaps I can override url helpers somewhere to sub out the `:path_key` based on `request.env`... but then how would I make it work in Mailers?

**Comment by janko** [comment]
The `:path_key` parameterized syntax is specific to the Rails router. It used to be supported by Roda, but that functionality has been extracted into a [plugin](https://roda.jeremyevans.net/rdoc/classes/Roda/RodaPlugins/PlaceholderStringMatchers.html). It wouldn't be possible to use it with Rodauth anyway, because `prefix` is used in `*_path` route helpers, which isn't connected to Roda routing.

You would dynamically evaluate the `prefix`:

```rb
class RodauthMain < Rodauth::Rails::Auth
  configure do
    prefix { "/api/auth/#{tenant_id}" }
    # ...
  end

  def tenant_id
    request.env[:tenant_id]
  end
end
```

You can then set the `request.env[:tenant_id]` in the mailer as described in [this comment](https://github.com/jeremyevans/rodauth/discussions/378#discussioncomment-7847792).

But I'd be interested to know how your Rails app is setting tenant ID. I would like to see if there was a more Railsy way of handling that in Rodauth. An alternative I see is adding a custom attribute accessor to the Rodauth object, and just setting the tenant ID there. For example, based on [your comment](https://github.com/jeremyevans/rodauth/discussions/378#discussioncomment-7857865):

```rb
class RodauthMain < Rodauth::Rails::Auth
  configure do
    prefix { "/api/auth/#{tenant.path_key}" }
    # ...
  end

  attr_accessor :tenant
end
```
```rb
class RodauthApp < Rodauth::Rails::App
  configure RodauthMain

  route do |r|
    r.on "api/auth", String do |path_key|
      tenant = Tenant.fetch_by_path_key(path_key)
      next unless tenant # consider route unmatched

      rodauth.tenant = tenant
      r.rodauth
    end
  end
end
```

You could then set the same in the mailer:

```rb
class RodauthMailer < ApplicationMailer
  # ...
  def rodauth(account_id, tenant_id)
    instance = RodauthApp.rodauth(name).allocate
    instance.tenant = Tenant.find(tenant_id)
    instance.instance_eval { @account = account_ds(account_id).first! }
    instance.instance_eval(&block) if block
    instance
  end
end
```

**Reply by pboling** [reply]
That's very helpful!  Thanks @janko :)

I don't understand one part though.

```
    r.on "api/auth", String do |path_key|
      tenant = Tenant.fetch_by_path_key(path_key)
      next unless tenant # consider route unmatched

      rodauth.tenant = tenant # <=========== Code is blowing up here, because ...
      r.rodauth
    end
```

I am setting this up for testing inside the existing test suite, and I've created a new Rodauth class called `class RodauthMultiTenant < Rodauth::Rails::Auth`, which is using the name `:multi_tenant`... but `rodauth` in that context is the instance of `RodauthMain`:

```
NoMethodError: undefined method `tenant=' for #<RodauthMain:0x000000012a773b40>
    app/misc/rodauth_app.rb:40:in `block (3 levels) in <class:RodauthApp>'
```

**Reply by pboling** [reply]
I am not sure how to set it on the specific Rodauth instane I need, but before it runs.

And if I remove the setting of the `tenant` on `rodauth`, there's a deeper issue: `KeyError` from the `route_hash` cache of path-key-less routes...
```
KeyError: key not found: "/multi/tenant/kiwi/create-account"
    /Volumes/drbl/src/vital/forks/rodauth-rails/lib/rodauth/rails/feature/base.rb:75:in `fetch'
    /Volumes/drbl/src/vital/forks/rodauth-rails/lib/rodauth/rails/feature/base.rb:75:in `rails_path_parameters'
    /Volumes/drbl/src/vital/forks/rodauth-rails/lib/rodauth/rails/feature/instrumentation.rb:38:in `rails_instrument_request'
    /Volumes/drbl/src/vital/forks/rodauth-rails/lib/rodauth/rails/feature/internal_request.rb:33:in `rails_instrument_request'
    /Volumes/drbl/src/vital/forks/rodauth-rails/lib/rodauth/rails/feature/instrumentation.rb:10:in `_around_rodauth'
    rodauth (2.32.0) lib/rodauth.rb:142:in `block (2 levels) in route'
```

I knew this was coming, and now I'm stuck again.  Any recommendations on the best way to get around the static keys of the `route_hash`?

**Reply by pboling** [reply]
I think I may have `path_key` set now:
```
    r.on("multi/tenant") do
      r.on(:path_key) do |path_key|
        # In a real life application you'd only proceed to rodauth routing if a tenant was found.
        # We'll mimic that possibility by skipping a /banana path-key
        next if path_key == "banana"

        r.env[:path_key] = path_key
        rodauth(:multi_tenant).path_key = path_key

        r.rodauth(:multi_tenant)
        next
      end
      next
    end
```

Continuing to try to figure out the route_hash issue, and it looks like your dynamic `prefix` is working, and I just need to remove some of the other hacks I had in place to allow it to work with the route_hash lookup!

**Reply by pboling** [reply, marked as answer]
💥 Boom! Got it working, added examples to test suite [here](https://github.com/pboling/rodauth-rails/pull/1).

**Reply by janko** [reply]
> Additionally - it appears that this whole bit won't work with dynamic paths. Is this going to prevent rodauth-rails from working?

I'm almost ready to close https://github.com/janko/rodauth-rails/pull/245 without merging. My main motivation was a simpler locale switcher, but I just opened https://github.com/jeremyevans/rodauth/pull/381, which will be a good enough solution. So, I advise that you don't base your implementation on that PR.

Glad you got the multi-tenancy working 👍🏻

**Reply by pboling** [reply]
rebasing my changes onto current `main` resulted in some test failures, but I'm not sure why.  Is the test suite working on `main`?

**Reply by pboling** [reply]
Yes, the tests are working... it sort of seems like my switch back to being based on latest `main` went badly...  Trying again.

**Reply by pboling** [reply]
OK, got it all working.  The branch is `multi-tenant` still, and I'll be using that for whatever I need to validate with multi-tenancy going forward.
